// vite-plugin-html/dist/index.mjs
var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};

// src/injectHtml.ts
import { render } from "ejs";
import path from "path";
import dotenv from "dotenv";
import dotenvExpand from "dotenv-expand";
import fs from "fs";
function injectHtml(options = {}) {
  let config;
  let env = {};
  return {
    name: "vite:inject-html",
    enforce: "pre",
    configResolved(resolvedConfig) {
      config = resolvedConfig;
      env = loadEnv(config.mode, config.root, "");
    },
    transformIndexHtml: {
      enforce: "pre",
      transform(html) {
        var _a, _b;
        const { injectData = {}, data, ejsOptions, injectOptions = {}, tags = [] } = options;
        const ejsData = __spreadValues(__spreadValues(__spreadValues(__spreadValues({}, (_a = config == null ? void 0 : config.env) != null ? _a : {}), (_b = config == null ? void 0 : config.define) != null ? _b : {}), env || {}), data || injectData);
        return {
          html: render(html, ejsData, ejsOptions || injectOptions),
          tags
        };
      }
    }
  };
}
function loadEnv(mode, envDir, prefix = "") {
  if (mode === "local") {
    throw new Error(`"local" cannot be used as a mode name because it conflicts with the .local postfix for .env files.`);
  }
  const env = {};
  const envFiles = [
    `.env.${mode}.local`,
    `.env.${mode}`,
    `.env.local`,
    `.env`
  ];
  for (const file of envFiles) {
    const path2 = lookupFile(envDir, [file], true);
    if (path2) {
      const parsed = dotenv.parse(fs.readFileSync(path2), {
        debug: !!process.env.DEBUG || void 0
      });
      dotenvExpand({
        parsed,
        ignoreProcessEnv: true
      });
      for (const [key, value] of Object.entries(parsed)) {
        if (key.startsWith(prefix) && env[key] === void 0) {
          env[key] = value;
        } else if (key === "NODE_ENV") {
          process.env.VITE_USER_NODE_ENV = value;
        }
      }
    }
  }
  return env;
}
function lookupFile(dir, formats, pathOnly = false) {
  for (const format of formats) {
    const fullPath = path.join(dir, format);
    if (fs.existsSync(fullPath) && fs.statSync(fullPath).isFile()) {
      return pathOnly ? fullPath : fs.readFileSync(fullPath, "utf-8");
    }
  }
  const parentDir = path.dirname(dir);
  if (parentDir !== dir) {
    return lookupFile(parentDir, formats, pathOnly);
  }
}

// src/minifyHtml.ts
import { minify } from "html-minifier-terser";
import { createFilter } from "@rollup/pluginutils";
function minifyHtml(minifyOptions = true) {
  const filter = createFilter(["**/*.html"]);
  const options = __spreadValues({
    collapseBooleanAttributes: true,
    collapseWhitespace: true,
    minifyCSS: true,
    minifyJS: true,
    minifyURLs: true,
    removeAttributeQuotes: true,
    removeComments: true,
    removeEmptyAttributes: true,
    html5: true,
    keepClosingSlash: true,
    removeRedundantAttributes: true,
    removeScriptTypeAttributes: true,
    removeStyleLinkTypeAttributes: true,
    useShortDoctype: true
  }, typeof minifyOptions === "boolean" ? {} : minifyOptions);
  return {
    name: "vite:minify-html",
    enforce: "post",
    async generateBundle(_options, outBundle) {
      if (options) {
        for (const bundle of Object.values(outBundle)) {
          if (bundle.type === "asset" && filter(bundle.fileName)) {
            bundle.source = minify(bundle.source, options);
          }
        }
      }
    }
  };
}

// src/index.ts
import { minify as minify2 } from "html-minifier-terser";
var minifyFn = minify2;
var src_default = (options = {}) => {
  const { inject = {}, minify: minify3 = {} } = options;
  return [injectHtml(inject), minifyHtml(minify3)];
};
export {
  src_default as default,
  injectHtml,
  minifyFn,
  minifyHtml
};
