// vite-plugin-html/dist/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};

// src/index.ts
__export(exports, {
  default: () => src_default,
  injectHtml: () => injectHtml,
  minifyFn: () => minifyFn,
  minifyHtml: () => minifyHtml
});

// src/injectHtml.ts
var import_ejs = __toModule(require("ejs"));
var import_path = __toModule(require("path"));
var import_dotenv = __toModule(require("dotenv"));
var import_dotenv_expand = __toModule(require("dotenv-expand"));
var import_fs = __toModule(require("fs"));
function injectHtml(options = {}) {
  let config;
  let env = {};
  return {
    name: "vite:inject-html",
    enforce: "pre",
    configResolved(resolvedConfig) {
      config = resolvedConfig;
      env = loadEnv(config.mode, config.root, "");
    },
    transformIndexHtml: {
      enforce: "pre",
      transform(html) {
        var _a, _b;
        const { injectData = {}, data, ejsOptions, injectOptions = {}, tags = [] } = options;
        const ejsData = __spreadValues(__spreadValues(__spreadValues(__spreadValues({}, (_a = config == null ? void 0 : config.env) != null ? _a : {}), (_b = config == null ? void 0 : config.define) != null ? _b : {}), env || {}), data || injectData);
        return {
          html: (0, import_ejs.render)(html, ejsData, ejsOptions || injectOptions),
          tags
        };
      }
    }
  };
}
function loadEnv(mode, envDir, prefix = "") {
  if (mode === "local") {
    throw new Error(`"local" cannot be used as a mode name because it conflicts with the .local postfix for .env files.`);
  }
  const env = {};
  const envFiles = [
    `.env.${mode}.local`,
    `.env.${mode}`,
    `.env.local`,
    `.env`
  ];
  for (const file of envFiles) {
    const path2 = lookupFile(envDir, [file], true);
    if (path2) {
      const parsed = import_dotenv.default.parse(import_fs.default.readFileSync(path2), {
        debug: !!process.env.DEBUG || void 0
      });
      (0, import_dotenv_expand.default)({
        parsed,
        ignoreProcessEnv: true
      });
      for (const [key, value] of Object.entries(parsed)) {
        if (key.startsWith(prefix) && env[key] === void 0) {
          env[key] = value;
        } else if (key === "NODE_ENV") {
          process.env.VITE_USER_NODE_ENV = value;
        }
      }
    }
  }
  return env;
}
function lookupFile(dir, formats, pathOnly = false) {
  for (const format of formats) {
    const fullPath = import_path.default.join(dir, format);
    if (import_fs.default.existsSync(fullPath) && import_fs.default.statSync(fullPath).isFile()) {
      return pathOnly ? fullPath : import_fs.default.readFileSync(fullPath, "utf-8");
    }
  }
  const parentDir = import_path.default.dirname(dir);
  if (parentDir !== dir) {
    return lookupFile(parentDir, formats, pathOnly);
  }
}

// src/minifyHtml.ts
var import_html_minifier_terser = __toModule(require("html-minifier-terser"));
var import_pluginutils = __toModule(require("@rollup/pluginutils"));
function minifyHtml(minifyOptions = true) {
  const filter = (0, import_pluginutils.createFilter)(["**/*.html"]);
  const options = __spreadValues({
    collapseBooleanAttributes: true,
    collapseWhitespace: true,
    minifyCSS: true,
    minifyJS: true,
    minifyURLs: true,
    removeAttributeQuotes: true,
    removeComments: true,
    removeEmptyAttributes: true,
    html5: true,
    keepClosingSlash: true,
    removeRedundantAttributes: true,
    removeScriptTypeAttributes: true,
    removeStyleLinkTypeAttributes: true,
    useShortDoctype: true
  }, typeof minifyOptions === "boolean" ? {} : minifyOptions);
  return {
    name: "vite:minify-html",
    enforce: "post",
    async generateBundle(_options, outBundle) {
      if (options) {
        for (const bundle of Object.values(outBundle)) {
          if (bundle.type === "asset" && filter(bundle.fileName)) {
            bundle.source = (0, import_html_minifier_terser.minify)(bundle.source, options);
          }
        }
      }
    }
  };
}

// src/index.ts
var import_html_minifier_terser2 = __toModule(require("html-minifier-terser"));
var minifyFn = import_html_minifier_terser2.minify;
var src_default = (options = {}) => {
  const { inject = {}, minify: minify3 = {} } = options;
  return [injectHtml(inject), minifyHtml(minify3)];
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  injectHtml,
  minifyFn,
  minifyHtml
});
